pkg/mimirpb/compat.go

```diff
@@ -214,11 +214,42 @@ func FromHistogramProtoToHistogram(hp *Histogram) *histogram.Histogram {
 }
 
 func FromHistogramProtoToFloatHistogram(hp *Histogram) *histogram.FloatHistogram {
+	if hp == nil {
+		return nil
+	}
+	if hp.IsFloatHistogram() {
+		panic("FromHistogramProtoToFloatHistogram called on float histogram")
+	}
+	return &histogram.FloatHistogram{
+		CounterResetHint: histogram.CounterResetHint(hp.ResetHint),
+		Schema:           hp.Schema,
+		ZeroThreshold:    hp.ZeroThreshold,
+		ZeroCount:        float64(hp.GetZeroCountInt()),
+		Count:            float64(hp.GetCountInt()),
+		Sum:              hp.Sum,
+		PositiveSpans:    fromSpansProtoToSpans(hp.GetPositiveSpans()),
```

Comment:
I know this PR doesn't add the `fromSpansProtoToSpans` method, but I'm looking at it with fresh eyes in this context... could it use an unsafe cast from `[]BucketSpan` to `[]histogram.Span` to avoid the need to allocate a new slice and populate it?

Reply:
> On the flip side I think it was you @charleskorn that added a cool way to unit test equivalence of structs via reflection, I'd apply that here as well

Yep, that was added [here](https://github.com/grafana/mimir/pull/4303/files#diff-a018202c8807d5f96df642a657b581faecaf6eaf7f0fcb38b18af018dd6bcbfc). The method I added there might need some tweaking to cover additional types, I only added what I needed for that situation.

Reply:
Expanded on in #4729

Reply:
Added issue #4742   so this PR can be merged